expr/func_movingwindow.go

```diff
@@ -0,0 +1,167 @@
+package expr
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/grafana/metrictank/batch"
+	"github.com/grafana/metrictank/errors"
+	"github.com/grafana/metrictank/schema"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/consolidation"
+	"github.com/raintank/dur"
+)
+
+type FuncMovingWindow struct {
+	in           GraphiteFunc
+	windowSize   string
+	fn           string
+	xFilesFactor float64
+
+	shiftOffset uint32
+}
+
+// NewMovingWindowConstructor takes an agg string and returns a constructor function
+func NewMovingWindowConstructor(name string) func() GraphiteFunc {
+	return func() GraphiteFunc {
+		return &FuncMovingWindow{fn: name}
+	}
+}
+
+func NewMovingWindow() GraphiteFunc {
+	return &FuncMovingWindow{fn: "average", xFilesFactor: 0}
+}
+
+func (s *FuncMovingWindow) Signature() ([]Arg, []Arg) {
+	return []Arg{
+			ArgSeriesList{val: &s.in},
+			ArgString{key: "windowSize", val: &s.windowSize, validator: []Validator{IsSignedIntervalString}},
+			ArgString{key: "func", opt: true, val: &s.fn, validator: []Validator{IsConsolFunc}},
+			ArgFloat{key: "xFilesFactor", opt: true, val: &s.xFilesFactor, validator: []Validator{WithinZeroOneInclusiveInterval}},
+		}, []Arg{
+			ArgSeriesList{},
+		}
+}
+
+func (s *FuncMovingWindow) Context(context Context) Context {
+	var err error
+	s.shiftOffset, err = s.getWindowSeconds()
+	if err != nil {
+		// shouldn't happen, validated above
+		log.Warnf("movingWindow: encountered error in getWindowSeconds, %s", err)
+		return context
+	}
+
+	// Adjust to fetch from the shifted context.from with
+	// context.to unchanged
+	context.from -= s.shiftOffset
+	return context
+}
+
+func (s *FuncMovingWindow) Exec(dataMap DataMap) ([]models.Series, error) {
+	series, err := s.in.Exec(dataMap)
+	if err != nil {
+		return nil, err
+	}
+
+	aggFunc := consolidation.GetAggFunc(consolidation.FromConsolidateBy(s.fn))
+	aggFuncName := fmt.Sprintf("moving%s", strings.Title(s.fn))
+	formatStr := fmt.Sprintf("%s(%%s,\"%s\")", aggFuncName, s.windowSize)
+
+	newName := func(oldName string) string {
+		return fmt.Sprintf(formatStr, oldName)
+	}
+
+	outputs := make([]models.Series, 0, len(series))
+	for _, serie := range series {
+
+		from, points, err := aggregateOnWindow(serie, aggFunc, s.shiftOffset, s.xFilesFactor)
+		if err != nil {
+			return nil, err
+		}
+
+		serie.Target = newName(serie.Target)
+		serie.QueryPatt = newName(serie.QueryPatt)
+		serie.Tags = serie.CopyTagsWith(aggFuncName, s.windowSize)
+		serie.QueryFrom = from
+		serie.Datapoints = points
+
+		outputs = append(outputs, serie)
+	}
+
+	dataMap.Add(Req{}, outputs...)
+	return outputs, nil
+}
+
+func aggregateOnWindow(serie models.Series, aggFunc batch.AggFunc, interval uint32, xFilesFactor float64) (uint32, []schema.Point, error) {
```

Comment:
windowsize, offset, interval. way too many terms for the same thing.

Reply:
Agreed that `offset` and `interval` are the same thing and refer to the windowSize in seconds. I can change this.
`windowSize` is the string argument passed in.